server/src/main/java/org/opensearch/common/xcontent/JsonToStringXContentParser.java

```diff
@@ -0,0 +1,257 @@
+import org.opensearch.common.xcontent.json.JsonXContent;
+import org.opensearch.core.xcontent.AbstractXContentParser;
+import org.opensearch.core.xcontent.DeprecationHandler;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.core.xcontent.XContentLocation;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.mapper.ParseContext;
+import java.io.IOException;
+import java.nio.CharBuffer;
+import java.util.ArrayList;
+
+/**
+ * JsonToStringParser is the main parser class to transform JSON into stringFields in a XContentParser
+ * returns XContentParser with one parent field and subfields
+ * fieldName, fieldName._value, fieldName._valueAndPath
+ * @opensearch.internal
+ */
+public class JsonToStringXContentParser extends AbstractXContentParser {
+    private final String fieldTypeName;
+    private XContentParser parser;
+
+    private ArrayList<String> valueList = new ArrayList<>();
+    private ArrayList<String> valueAndPathList = new ArrayList<>();
+    private ArrayList<String> keyList = new ArrayList<>();
+
+    private XContentBuilder builder = XContentBuilder.builder(JsonXContent.jsonXContent);
+    private ParseContext parseContext;
+
+    private NamedXContentRegistry xContentRegistry;
+
+    private DeprecationHandler deprecationHandler;
+
+    private static final String VALUE_AND_PATH_SUFFIX = "._valueAndPath";
+    private static final String VALUE_SUFFIX = "._value";
+    private static final String DOT_SYMBOL = ".";
+    private static final String EQUAL_SYMBOL = "=";
+
+    public JsonToStringXContentParser(
+        NamedXContentRegistry xContentRegistry,
+        DeprecationHandler deprecationHandler,
+        ParseContext parseContext,
+        String fieldTypeName
+    ) throws IOException {
+        super(xContentRegistry, deprecationHandler);
+        this.parseContext = parseContext;
+        this.deprecationHandler = deprecationHandler;
+        this.xContentRegistry = xContentRegistry;
+        this.parser = parseContext.parser();
+        this.fieldTypeName = fieldTypeName;
+    }
+
+    public XContentParser parseObject() throws IOException {
+        builder.startObject();
+        StringBuilder path = new StringBuilder(fieldTypeName);
+        parseToken(path, null);
+        builder.field(this.fieldTypeName, keyList);
+        builder.field(this.fieldTypeName + VALUE_SUFFIX, valueList);
+        builder.field(this.fieldTypeName + VALUE_AND_PATH_SUFFIX, valueAndPathList);
+        builder.endObject();
+        String jString = XContentHelper.convertToJson(BytesReference.bytes(builder), false, XContentType.JSON);
+        return JsonXContent.jsonXContent.createParser(this.xContentRegistry, this.deprecationHandler, String.valueOf(jString));
+    }
+
+    private void parseToken(StringBuilder path, String currentFieldName) throws IOException {
+
+        while (this.parser.nextToken() != Token.END_OBJECT) {
```

Comment:
Question here please, there are a limits on mappings that OpenSearch imposes on index level [1], should those be taken into account for flat objects:
 - `index.mapping.depth.limit` - the depth of JSON object structure
 - `index.mapping.field_name_length.limit` - the name of the flattened field 

I suspect that depth limit would probably make sense to respect.

[1] https://www.elastic.co/guide/en/elasticsearch/reference/7.17/mapping-settings-limit.html

Reply:
In this first PR, I didn't open any parameters and set default parameters. If we are planning to open up parameters, there will be wider discussion, for example, depth_limit, fieldname_length_limit, value_length_limit, split_queries_on_whitespace, doc_values(able to allow users set it to false). Shall we develop the open parameters in the next PR maybe in 2.8？

Reply:
> Shall we develop the open parameters in the next PR maybe in 2.8？

I think it makes sense, we could refine that later